File: clients/python/examples/rankings.py
"""Rank aggregation: combine best-first ballots into one ranking.

Run after `maturin develop`:  python examples/rankings.py
"""

import propagon


def main() -> None:
    ballots = propagon.RankingsDataset()
    for ballot in [
        ["espresso", "latte", "drip", "tea"],
        ["espresso", "drip", "latte", "tea"],
        ["latte", "espresso", "tea", "drip"],
        ["drip", "espresso", "latte", "tea"],
    ]:
        ballots.push_ranking(ballot)

    print("Plackett-Luce strengths:")
    for name, score in propagon.PlackettLuce().fit(ballots).sorted_scores():
        print(f"  {name:9s} {score:.4f}")

    # Kemeny / Borda work on pairwise data; lower ballots to ordered pairs.
    pairs = propagon.PairwiseDataset()
    for ballot in [
        ["espresso", "latte", "drip", "tea"],
        ["espresso", "drip", "latte", "tea"],
        ["latte", "espresso", "tea", "drip"],
        ["drip", "espresso", "latte", "tea"],
    ]:
        for i in range(len(ballot)):
            for j in range(i + 1, len(ballot)):
                pairs.push(ballot[i], ballot[j])

    print("\nBorda count:")
    for name, score in propagon.Borda().fit(pairs).sorted_scores():
        print(f"  {name:9s} {score:.0f}")


if __name__ == "__main__":
    main()

File: clients/wasm/examples/rankings.ts
// Rank aggregation: Plackett-Luce over ballots, Borda over pairwise data.
//   node examples/rankings.ts
import { datasets, rankings, pairwise } from "../index.js";

const ballots = [
  ["espresso", "latte", "drip", "cold-brew"],
  ["espresso", "drip", "latte", "cold-brew"],
  ["latte", "espresso", "cold-brew", "drip"],
  ["espresso", "latte", "cold-brew", "drip"],
];

const r = new datasets.RankingsDataset();
for (const ballot of ballots) r.pushRanking(ballot);

console.log("Plackett-Luce:");
for (const [name, score] of rankings.fitPlackettLuce({}, r).sortedScores()) {
  console.log(`  ${name.padEnd(10)} ${score.toFixed(4)}`);
}

// Borda over pairwise preferences extracted from the same ballots.
const p = new datasets.PairwiseDataset();
for (const ballot of ballots) {
  for (let i = 0; i < ballot.length; i++) {
    for (let j = i + 1; j < ballot.length; j++) p.push(ballot[i], ballot[j], 1);
  }
}
console.log("\nBorda (from pairwise):");
for (const [name, score] of pairwise.fitBorda(p).sortedScores()) {
  console.log(`  ${name.padEnd(10)} ${score.toFixed(4)}`);
}
